pkg/csi/service/fcd/controller_test.go

```diff
@@ -0,0 +1,644 @@
+			t.Errorf("[CREATE] Name of FCD does not match %s != %s", volName, volNameRet)
+		}
+	}
+
+	// order should be repeatable
+	// first call
+	respFirst, err := c.ListVolumes(ctx, &csi.ListVolumesRequest{})
+	if err != nil {
+		t.Errorf("[FIRST] ListVolumes failed: %v", err)
+	}
+	countFirst := len(respFirst.Entries)
+	if countFirst != 11 {
+		t.Errorf("There should 11 FCD present count=%d", countFirst)
+	}
+
+	// second call
+	respSecond, err := c.ListVolumes(ctx, &csi.ListVolumesRequest{})
+	if err != nil {
+		t.Errorf("[SECOND] ListVolumes failed: %v", err)
+	}
+	countSecond := len(respSecond.Entries)
+	if countFirst != 11 {
+		t.Errorf("There should 11 FCD present count=%d", countSecond)
+	}
+
+	for i := 0; i < 11; i++ {
+		firstName := respFirst.Entries[i].Volume.VolumeContext[AttributeFirstClassDiskName]
+		secondName := respSecond.Entries[i].Volume.VolumeContext[AttributeFirstClassDiskName]
+
+		if !strings.EqualFold(firstName, secondName) {
+			t.Errorf("FirstName(%s) != SecondName(%s)", firstName, secondName)
+		}
+	}
+}
+
+func TestZoneSupport(t *testing.T) {
+	// context
+	ctx := context.Background()
+
+	// Start vcsim
+	config, cleanup := configFromEnvOrSim(true)
+	defer cleanup()
+
+	connMgr := cm.NewConnectionManager(&config, nil)
+	defer connMgr.Logout()
+
+	c := &controller{
+		cfg:     &config,
+		connMgr: connMgr,
+	}
+
+	err := connMgr.Connect(ctx, config.Global.VCenterIP)
+	if err != nil {
+		t.Errorf("Failed to Connect to vSphere: %s", err)
+	}
+
+	// Get the vSphere Instance
+	vsi := connMgr.VsphereInstanceMap[config.Global.VCenterIP]
+
+	// Tag manager instance
+	restClient := rest.NewClient(vsi.Conn.Client)
+	user := url.UserPassword(vsi.Conn.Username, vsi.Conn.Password)
+	if err := restClient.Login(ctx, user); err != nil {
+		t.Fatalf("Rest login failed. err=%v", err)
+	}
+
+	m := tags.NewManager(restClient)
+
+	/*
+	 * START SETUP
+	 */
+	// Create a region category
+	regionID, err := m.CreateCategory(ctx, &tags.Category{Name: config.Labels.Region})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create a region tag
+	regionID, err = m.CreateTag(ctx, &tags.Tag{CategoryID: regionID, Name: "k8s-region-US"})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create a zone category
+	zoneID, err := m.CreateCategory(ctx, &tags.Category{Name: config.Labels.Zone})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create a zone tags
+	zoneIDwest, err := m.CreateTag(ctx, &tags.Tag{CategoryID: zoneID, Name: "k8s-zone-US-west"})
+	if err != nil {
+		t.Fatal(err)
+	}
+	zoneIDeast, err := m.CreateTag(ctx, &tags.Tag{CategoryID: zoneID, Name: "k8s-zone-US-east"})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Setup a multi-DC environment with zones!
+	// Setup DC0
+	dc0, err := vclib.GetDatacenter(ctx, vsi.Conn, "DC0")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Attach a random tag to DC0
+	if err = m.AttachTag(ctx, regionID, dc0.Reference()); err != nil {
+		t.Fatal(err)
+	}
+	if err = m.AttachTag(ctx, zoneIDwest, dc0.Reference()); err != nil {
```

Comment:
No need to call `.Reference()` in any of the AttachTag calls, the param is a `mo.Reference` interface, which all the inventory types satisfy.

Reply:
Got it.